### PR TITLE
Self loop asynchronous v2

### PR DESCRIPTION
**Pull request by jameshalgren**
@hellkite500, @jhrehanoaa, @groutr -- this is the async long-run simulation with the warmstates and  output added. The goal is for the following structure:

image

```
        '''
        Asynchronous execution Psuedocode
        Sync1: Prepare first warmstate from files
        Sync1: Prepare first forcing from files

        For first forcing set
            Sync2a: run model
            Sync2b: begin preparing next forcing
            Sync3a - AFTER Sync2a, prepare next warmstate (last state of model run)
            Sync3b: write any output from Sync2a
            Loop has to wait for Sync2a+b+Sync3a, does not have to wait for Sync3b
                  if next forcing prepared
        '''
```

The following script (based first [on this gist](https://gist.github.com/groutr/7f09488220de622ccfa40c126e91e15c
)) provides a useful demonstration of how this can be accomplished with functions that would otherwise block truly concurrent threading: 

``` py
print("Requires python 3.8+")
print("arg1 -- number of loops")
print("arg2 -- 1/0 use asyncio or not -- currently makes no difference, it seems")
print("arg3 -- 1/0 use CPU-heavy model imitator or not")
print("arg4 -- 0/1 force (0 allows separate threads) all threads to the main kernel thread")

'''
Useful references:
https://gist.github.com/groutr/7f09488220de622ccfa40c126e91e15c
https://www.joeltok.com/blog/2021-2/python-async-sync
https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread
https://trio.readthedocs.io/en/stable/design.html
https://curio.readthedocs.io/en/latest/
'''

import asyncio
import time
import random
import hashlib
import sys
import concurrent.futures
import os

is_async = bool(int(sys.argv[2]))
use_cpu = bool(int(sys.argv[3]))
force_kernel_thread = bool(int(sys.argv[4]))

# create_thread_pool_here
if force_kernel_thread:
    pool_IO = None
    pool_Processing = None
else:
    pool_IO = concurrent.futures.ThreadPoolExecutor(max_workers=4)
    pool_Processing = concurrent.futures.ThreadPoolExecutor(max_workers=4)

longrun = 6
mediumrun = 2
shortrun = 1

def warm_state(t, loop):
    print(f"{loop}: Starting to prepare warm state...", t)
    rv = random.getrandbits(64)
    time.sleep(t)
    print(f"{loop}: Finished preparing warm state")
    return rv

def read_forcing(t, loop):
    print(f"{loop}: Beginning to read forcings...", t)
    rv = random.getrandbits(64)
    time.sleep(t)
    print(f"{loop}: Finished reading forcings")
    return rv

def run_model_sleep(t, loop, w, f):
    print(f"{loop}: Running model...", t)
    print("  Warm bytes:", w)
    print("  Forcing bytes:", f)
    time.sleep(t)
    print(f"{loop}: Finished running model")
    return ["this is the result"]

def run_model_CPU(t, loop, w, f):
    print(f"{loop}: Running model...", t)
    print("  Warm bytes:", w)
    print("  Forcing bytes:", f)
    start_time = time.perf_counter()
    m = hashlib.sha512()
    while time.perf_counter() - start_time < t:
        m.update(os.urandom(4098))
    print(f"{loop}: Finished running model")
    return ["this is the result"]

def write_output(t, loop, result):
    print(f"{loop}: Writing output...", t)
    print(f"{loop}: Writing {result}")
    time.sleep(t)
    print(f"{loop}: Finished writing output")

async def main(loops):

    if use_cpu:
        run_model = run_model_CPU
    else:
        run_model = run_model_sleep

    loop = asyncio.get_running_loop()

    forcings = loop.run_in_executor(pool_IO, read_forcing, mediumrun, -1)
    w = warm_state(shortrun, -1)

    for i in range(loops - 1):
        f = await forcings
        mtask = loop.run_in_executor(pool_Processing, run_model, longrun, i, w, f)

        forcings = loop.run_in_executor(pool_IO, read_forcing, mediumrun, i)

        m = await mtask
        w = warm_state(shortrun, i)
        write = loop.run_in_executor(pool_IO, write_output, longrun, i, m)


    i += 1
    f = await forcings
    mtask = loop.run_in_executor(pool_Processing, run_model, longrun, i, w, f)
    m = await mtask
    write = await loop.run_in_executor(pool_IO, write_output, longrun, i, m)


if __name__ == "__main__":
    loops = int(sys.argv[1])

    start_time = time.perf_counter()

    coroutine = main(loops)
    loop = asyncio.get_event_loop()

    if is_async:
        asyncio.run(coroutine)
    else:
        # This was a test to see if run_until_complete would force
        # serial execution on underlying calls... it doesn't, at
        # least in our example here.
        loop.run_until_complete(coroutine)

    print("Elapsed Time", time.perf_counter() - start_time)
```

### CODE REVIEW

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 1]
```diff
@@ -1586,10 +1588,25 @@ def new_nwm_q0(run_results):
                 debuglevel,
             )
 
-            #q0 = run_results
-            q0 = new_nwm_q0(run_results)
+        loop.run_until_complete(asyncio.gather(execute_model, build_next_forcing))
+        run_results = execute_model.result()  # gives us whatever is returned, once it is done.
+        qlats, usgs_df = build_next_forcing.result()  # gives us whatever is returned, once it is done.
+
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: Lines 1591-1594 appear to be a merge error; they should be superseded by the similar lines under the if statement starting on line 1598.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 2]
```diff
@@ -1586,10 +1588,25 @@ def new_nwm_q0(run_results):
...
+        #loop.run_until_complete(build_warm_state)
+        #loop.run: output
+        # TODO: This may create the last forcing twice -- Fix it.
+        if run_set_iterator < len(run_sets) - 1:
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: Let's test this logic and flow with some simple sleep/print commands...

Comment (jameshalgren) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: @groutr -- thanks for the suggestion!

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 3]
```diff
@@ -1586,10 +1588,25 @@ def new_nwm_q0(run_results):
...
+            qlats, usgs_df = build_next_forcing.result()  # gives us whatever is returned, once it is done.
+            new_q0 = loop.run_in_executor(None, new_nwm_q0, run_results)
+            loop.run_until_complete(asyncio.gather(new_q0))
+            q0 = new_q0.result()
+        else:
+            loop.run_until_complete(asyncio.gather(execute_model))
+            run_results = execute_model.result()  # gives us whatever is returned, once it is done.
 
-        nwm_output_generator(
+        output_last_run = loop.run_in_executor(None, nwm_output_generator,
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:1605]: If writing is to distinct files, then asynchronous is no problem. If writing is to the same outputfile (i.e. appending) then we have a race condition here.

**Comment by jameshalgren** [comment]
@groutr, @hellkite500 --- I added a few additional command line options: one to toggle between a CPU-intensive "model" and one that is not; and another to force everything onto the main thread.

**Comment by groutr** [comment]
@jameshalgren it's probably time to create an argparse commandline interface.

**Comment by jameshalgren** [comment]
Agreed (if that isn't overkill for this little demo). Might not hurt to get this into a real gist, too or built as a unit test somehow directly in the repo. I had also tried to figure out how it would have to work with python 3.6 -- still not there with that yet.

**Comment by jameshalgren** [comment]
ping @awlostowski-noaa

**Comment by jameshalgren** [comment]
Closing -- replaced by https://github.com/NOAA-OWP/t-route/pull/443

**Pull request closed without merging**